cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java

```diff
@@ -370,6 +372,22 @@ void addApplication(Id.Application id,
   WorkflowToken getWorkflowToken(Id.Workflow workflowId, String workflowRunId);
 
   /**
+   * Add node state for the given {@link Workflow} run. This method is used to update the
+   * state of the custom actions started by Workflow.
```

Comment:
`WorkflowNodeStateDetail` instance simply contains the node name, node status, run id and Throwable. So it can be created in the `AppMetaStore` using information passed in setStart and setStop method parameters.

Corresponding code is
```java
WorkflowNodeThrowable defaultThrowable = failureCause == null ? null : new WorkflowNodeThrowable(failureCause);
WorkflowNodeStateDetail nodeStateDetail = new WorkflowNodeStateDetail(workflowNodeId,
                                                                          ProgramRunStatus.toNodeStatus(status),
                                                                          pid, defaultThrowable);

```